File: capibara-website/app/dtos/RefResult.scala
package dtos
import models._
import play.api.libs.json.Json

object RefResult {
  implicit val refResultFormatter = Json.format[RefResult];
}

case class RefResult(
    header: Option[HeaderDetail] = None,
    `macro`: Option[Macro] = None,
    `enum`: Option[Enum] = None,
    struct: Option[Struct] = None,
    typedef: Option[Typedef] = None,
    function: Option[Function] = None
) {
  require(
    header.isDefined ^ `macro`.isDefined ^ `enum`.isDefined ^ struct.isDefined ^ typedef.isDefined ^ function.isDefined,
    "Exactly one of header, macro, enum, struct, typedef, and function must exclusively be defined!"
  );
}

File: capibara-website/src/dtos/ref_result.rs
use rocket::serde::{Deserialize, Serialize};

use crate::dtos::header_detail::HeaderDetail;
use crate::models::_macro::Macro;
use crate::models::enumeration::Enumeration;
use crate::models::_struct::Struct;
use crate::models::typedef::Typedef;
use crate::models::function::Function;

#[derive(Serialize, Deserialize, Clone, Default)]
pub(crate) struct RefResult {
    pub header: Option<HeaderDetail>,
    #[serde(rename = "macro")]
    pub _macro: Option<Macro>,
    #[serde(rename = "enum")]
    pub _enum: Option<Enumeration>,
    #[serde(rename = "struct")]
    pub _struct: Option<Struct>,
    pub typedef: Option<Typedef>,
    pub function: Option<Function>
}

impl RefResult {
    pub fn is_returnable(&self) -> bool {
        self.header.is_some() ||
        self._macro.is_some() ||
        self._enum.is_some() ||
        self._struct.is_some() ||
        self.typedef.is_some() ||
        self.function.is_some()
    }
}